### ISSUE TITLE
consider to switch to zsh

### ISSUE THREAD

**Issue by philmmanjaro** [manjaro/release-plan#47]
On further improving shell experience and colors and stuff we can switch to zsh. It's compatible with bash, but much faster.

It will take some work to deploy a new config for zsh. but... there's **oh-my-zsh** for that, which ships a bucketload of themes. It will also improve our development work as **zsh** would for example instantly display git branch status and so on. So it is the very same thing, and even with additional functionality. Arch already uses zsh as default shell on their live-media.

**Comment by philmmanjaro** [comment]
Here is an [oldy but goldy](http://teom.org/blog/kde/color-your-shell/) from Teo about this topic.

**Comment by udeved** [comment]
That wouldn't hurt at all, because upstream zsh package does not deploy any /etc/skel files.
So the entire bash config with extension files became obsolete, which is caused by bash upstream providing /etc/skel/ bashrc files

**Comment by udeved** [comment]
https://github.com/manjaro/manjaro-tools/commit/5e1246d638d9eeb505431aa1d2a18c7ba49b255b

**Comment by Chrysostomus** [comment]
I think switching to zsh is a good idea. I would suggest that we ship simple zshrc instead of oh-my-zsh. Something with prompt similar to our bash prompt (or simpler), and autocd enabled. Oh-my-zsh is too heavy and complex for most users.

**Comment by philmmanjaro** [comment]
**CAL** supports this now with the **development** version. However the new **users** module still needs a better UI.

**Comment by jonathonf** [comment]
ZSH isn't a drop-in BASH replacement; I'd be very cautious about assuming things would "just work".

Most scripts are written to target a portable `/bin/sh` or, when extra functions are needed, `/bin/bash`. Noone writes portable scripts targeted at `/usr/bin/zsh`. Therefore BASH would need to be installed by default alongside ZSH, and in that case I can't see the benefit of setting ZSH as the default as a user can switch their shell as they wish.

If you can list any distros that have a default shell other than BASH (or DASH, but that's Ubuntu so meh) I'll reconsider. :)

**Comment by Chrysostomus** [comment]
I think arch and gentoo use it on their install media. I could be wrong though. 

Not having bash is not an option, because so many basic tools depend on it. 

Sure, for scripting purposes zsh is not a drop in replacement (for example anything using arrays should break). But as a login shell it definitely is (properly configured). I don't think anyone is suggesting we replace /bin/sh with it, 'cause that would not make any sense. Switching to dash might yield some performance increase, but with zsh the benefit would be negligible. 

But as a login shell it would definitely make sense. It is more user friendly, which would be in spirit of manjaro. And unlike fish, it doesn't break anything.

**Comment by Chrysostomus** [comment]
I propose we make a configuration package for zsh. zsh-settings, manjaro-zshrc or something. I suggest something similar to what bspwm edition has, but with prompt closer to bash default.

**Comment by Chrysostomus** [comment]
I made a pull request at packages-community, adding manjaro-zsh-config. 

Features:
- bash-style (but green) prompt 
- ctrl+arrows moves cursor a word, ctrl+backspace deletes previous word (xterm style movement) 
- autocompletion and smart autocorrect 
- fish style autosuggestion, syntax highlighting and history substring search 
- sets window title based on command 
- colored manpages 
- set base16 color themes, many options and easy configuration 
- set editor to nano 

I made this with the intention of offering it for common manjaro zsh configuration. It should similar enough to bash to be approachable, and well enough commented to be understandable, while improving shell experience for beginners and advanced users alike. 

The files are here: https://github.com/Chrysostomus/manjaro-zsh-config

Maybe we could use this as a starting point for finding a configuration that would be agreeable to most of us?

**Comment by excalibur1234** [comment]
i like your .zshrc file.

i have just tested it and the only thing i would change are:
- zsh-autosuggestions is listed as dependency in PKGBUILD twice
- HISTSIZE and savehist are really small. i would suggest 1000.
- the prompt should be red after going root with "su"

**Comment by Chrysostomus** [comment]
Thanks, I'll update it. Imho 100 is enough because no duplicates are saved, but 1000 is fine too and more common.

**Comment by Chrysostomus** [comment]
Hmm... I think I found a way to improve word recognition (as in don't consider full file paths single words for the purposes of movement and word deleting).  I think I'll add that

**Comment by excalibur1234** [comment]
the first line looking for /etc/profile.d/cnf.sh only works, if pkgfile is installed:
https://wiki.archlinux.org/index.php/Pkgfile
https://wiki.archlinux.org/index.php/Zsh#The_.22command_not_found.22_hook
at least this is what i have read about it. but i have not tested that. 
i am also not sure, whether pkgfile can work in manjaro or not

please make it work (e.g. by making pkgfile a dependency) or remove that line.

**Comment by Chrysostomus** [comment]
It is there to be ready when we get support for .files. https://github.com/manjaro/release-plan/issues/98 No point in adding pkgfile before it works, but once it does, we can just add the dependency. But this preparation for it is harmless.

**Comment by excalibur1234** [comment]
- do you know of any custom prompts for git repositories? these are really popular in oh-my-zsh. this is one of the reason philm mentioned in the first post.
- there is still the problem with "su". do you think the prompt should become red when in "root mode"? or should it stay the way it is?
i think we would need a separate .zshrc file for root if we want to do that.

**Comment by Chrysostomus** [comment]
Currently the $ becomes # when shell has elevated privileges. I can modify it to change color too. 

Maybe we could conditionally display git status in right prompt when in git repo? I have used some git status prompts, what information should be displayed?

**Comment by jonathonf** [comment]
Just a note of caution - be careful not to reinvent the wheel. There are things like `powerline` that will do an awful lot of what you're talking about without a great deal of extra configuration.

All you'd probably need is to change the default blue prompt to a Manjaro green.

**Comment by Chrysostomus** [comment]
Yes, I'm not going to write anything from scratch, for reinvented wheels are expensive and often full of angles. 

I think powerline is a bit much for average user, I would leave that to user discretion. 

This right prompt would be unintrusive, as it is displayed only when in git repos: https://techanic.net/2012/12/30/my_git_prompt_for_zsh.html

**Comment by excalibur1234** [comment]
what information should be displayed?

the branch name or commit hash ( if in detached head state) you are working on and maybe, the git status (whether all changes are committed or not). these are the things i have noticed the most using the git themes in oh-my-zsh.

i have always looked at [powerline](https://github.com/powerline/powerline) as an overkill, especially for non-vim users.

**Comment by archderwelt** [comment]
My two cents :    
Making zsh default shell doesn't/won't "replace" bash, in fact scripts can still be run with bash (#!/bin/bash) in scripts or bash -c "command" in command line.     
One of the most benefits of zsh is auto completion and  from my experience without grml-zsh-config package installed you get half the power of zsh IMO.

**Comment by Chrysostomus** [comment]
Would you then suggest that we enable menu completion like grml does?

**Comment by excalibur1234** [comment]
i have just uploaded my customized version of your .zshrc file:
https://github.com/excalibur1234/manjaro-zsh-config/blob/master/.zshrc

i have also adopted the git prompt from your link above and included the display of exit code from last command (look at lines 132-208)

**Comment by Chrysostomus** [comment]
Looks interesting, I like the cd aliases and the fact that it lists contents when entering a new directory. 

I don't think we should alias pacui and yaourt if we aim this to be common for all editions, because not all editions have it. For a similar reason I refrained from adding a keybinding to bmenu... 

I'll adopt most of your modifications, I think.

**Comment by jonathonf** [comment]
Please don't override built-in commands. That's going to get annoying very quickly. `cd` means "change directory", nothing else. ;)

**Comment by Chrysostomus** [comment]
Yeah, that part did not get merged. It's fine for personal use but not as common default.

**Comment by jonathonf** [comment]
I was going to edit but you got back too quickly. :P I should have said, adding an alias e.g. `cdl` which does `cd $0; ls;` would be fine.

**Comment by excalibur1234** [comment]
as you have seen, this is my personal .zshrc file.
this means that there are personal things in it, which are not recommended for general use.

this includes most aliases.
the "play" function looks for music files in your current directory (or the directory you have specified and plays it in a playlist with mpv. if a web address is given, it tries to play the music there.
this function is not meant for general use, because it depends on mpv!

also the "cd && ll" function is really useful in bash, but not really necessary in zsh. i recommend to remove it from there (or at least uncomment it by default.

thanks for reminding me about the "sudo" alias. i just removed it, too.

**Comment by Chrysostomus** [comment]
I have no idea what that sudo thing does, I think I got it from somewhere. Pacui and yaourt aliases were not merged. 

Merged features currently:
- autocd 
- completion (no menu, just good tab completion) 
- fish style autosuggest, syntax highlighting and history substring search 
- colored man pages 
- xterm style movement, / is not considered part of a word so whole file paths are not deleted at once 
- base16 colors theme switcher
- right prompt: hidden by default, reports exit status code if not 0, gives git status if in git repo
- set window title 
- some misc stuff

**Comment by jonathonf** [comment]
> as you have seen, this is my personal .zshrc file.
> this means that there are personal things in it, which are not recommended for general use.

That's fine - just you linked to it in this thread so I thought it was related.

**Comment by excalibur1234** [comment]
i have taken a look at grml-zsh-config as suggested above.
but the problem is that grml-zshrc is REALLY customized. you need a while to become familiar with it - and its syntax cannot simpy copied to an ordinary .zshrc file.

either way, i think we already have a lot of features in our .zshrc file already.

**Comment by Chrysostomus** [comment]
I actually started by trying to modify grml-zsh-config, but soon gave up because of its complexity. 

I think that if we keep adding much more features, the shell startup time is going to suffer.

**Comment by excalibur1234** [comment]
i agree. our zsh should be more comfortable than default bash while not being noticeably slower - even on slow computers.

if there are are no obvious features missing, i recommend to push manjaro-zsh-config into the unstable branch and make an announcement (including a description how to change shell to zsh).
then, people can start to test it and give feedback. 
we can also ask the rest of the team for feedback.

**Comment by archderwelt** [comment]
@Chrysostomus  I don't think it "needs" to be modified, it works(TM) out of the box and  offers a basic git prompt. That's what the archlinux install media has, zsh + grml-zsh-config and it is more than enough for installing and system administration.

**Comment by Chrysostomus** [comment]
@archderwelt Thanks for feedback! Manjaro-architect now offers choice of grml-zsh-config and manjaro-zsh-config after installing a manjaro edition. 

I would prefer that we would prefer our own configuration because 
- branding. There is nothing manjaro specific about grml-zsh-config. It suits arch well, because the default prompt uses arch colors (blue) and arch has philosophy of shipping unmodified upstream packages. Manjaro does not have the same ideal and brands software in general to fit the overall look and feel of the desktop. 
- comparing manjaro-zsh-config and grml-zsh-config, former provides extra features that benefit beginners and average users, while latter provides extra features mainly useful for system administrators. Manjaro version provides syntax highlighting and autosuggestion, while grml provides stuff like keybinding to insert current time. 
- manjaro-zsh-config is more discoverable for new user. The function comes from a single well commented file and the sourced plugins are easy to find. Imho grml-zsh-config is quite a bit more mysterious about where it keeps its files, and the configuration file is long and difficult to understand. 

I started manjaro-zsh-config, because grml-zsh-config just doesn't feel very manjaro to me, but more like arch/debian. That being said, grml is still a valid and respectable choice.

**Comment by archderwelt** [comment]
@Chrysostomus  True, the distros have different ideals and concerning the point of branding/customizing ["user friendly" <=> "do it your self/keep it simple and stupid"]. Also true grm-zsh-config has 'nothing' manjaro specific, but [nothing arch specific](https://git.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/grml-zsh-config) as well, the blue color is just by chance. Am using zsh/grml combo on my manjaro (stable) machine and it rocks :-) . I will have to try the manjaro version, so far I can't say any thing yet about it.    

> a bit more mysterious about where it keeps its files    

`pacman -Ql grml-zsh-config` should "print" the files.

**Comment by Chrysostomus** [comment]
Thanks, I do know how to find it, I was using colorful language to illustrate their difference in complexity. manjaro-zsh-config is just the zshrc with scripts that are visibly sourced. 

By arch specific I meant that it is congruent with the arch way to ship upstream packages. It is actual goal with arch linux, while manjaro is indifferent to the matter. Some packages are upstream, some are manjaro branded.

**Comment by Chrysostomus** [comment]
Could you further tell what is it that is like about grml-zsh-config? You obviously have been using it longer than I. 

Things I liked: 
- you can set the prompt theme by running 

     prompt <name>
- colored man pages. This idea I copied 
- light, no startup delay is introduced 

I have mixed feelings about it using a menu for tab completion. It looks really cool, but in practice I prefer normal tab completion.

**Comment by archderwelt** [comment]
While I am a typical "standard" user, I can't dig in technical things and all I can say for the moment is :    
I have used zsh without grm-zsh-config and the user experience is quite different, e.g if a file/dir name is uppercase and you type it lowercase, it feels like bash, but with grml-zsh it doesn't matter. You type first letter(s) lowercase, hit tab and name gets corrected. A big plus for me, saves pressing shift.    
Apart from the enhanced auto completion, it sets also sane defaults and plenty of helpful stuff in /etc/zsh/zshrc (line 1138 for example), these features might be overkill but a starting point for customizing and adjusting to fit the manjaro ideals.    
It feels like the two completes each other, this is why I suggested it.

**Comment by Chrysostomus** [comment]
Thanks for input. Manjaro-zsh-config has case insensitive completion enabled too. I'll look into grml further to see if I can find something interesting.

**Comment by excalibur1234** [comment]
i still think that .zshrc should also be copied into the /root/ directory.
this way, the command prompt becomes red after typing "su" and giving your password. without .zshrc in /root/ directory, you simply see a default shell.

is there any reason for not doing this?

**Comment by Chrysostomus** [comment]
I haven't gotten around to do it yet.

**Comment by Chrysostomus** [comment]
Done now in 0.2 update. Also added undo (shift+tab), switched prompt to maia and made some other tweaks.

**Comment by jonathonf** [comment]
I'd be wary of having giving root a heavily-customised profile (for any shell). For example, if it sources a configuration file it is doing so as root. Some of the options may also not be appropriate - for example, `option nohup` sounds like it could be pretty awful if you're not aware of what it does, and `option autocorrect` is not something that should be enabled for root. If you mess up your system because you mistype, that's one thing. If "autocorrect" gets it wrong, that's unforgivable (plus what you actually typed won't show up in history).

I'd err on the side of keeping root's shell as a plain as possible. Anyone messing round as root can always introduce their own configuration - and if you're operating in a root shell this much you're doing it wrong.

**Comment by udeved** [comment]
I see root account config for apps, besides some colours in terminal, very critical. I am no fan, to provide eg thunar root settings. Users should be reminded they are using root power, ugly theme is easiest reminder.

**Comment by oberon2007** [comment]
I think actually what it does is provide config to make root shell look different from normal user shell. Or am I wrong, @Chrysostomus ?

**Comment by udeved** [comment]
I am generally no fan of over customization.
If a user wants to have an alias to terminate a process, he should add it. 
This kind of stuff should not be default, we also do some not so good stuff with bashrc.

**Comment by Chrysostomus** [comment]
Good points about root user. I'll make adjustments.

**Comment by Chrysostomus** [comment]
I now made a separate, more plain zshrc for root. I doesn't source anything, does not have nohub set and no autocorrect. The only alias I kept for root is `alias cp="cp -i"`. It still has good tab completion, ctrl+arrows movement, colored man pages and autocd. The prompt is now exact bash prompt replica. 

For normal user I also kept  `alias x='startx ~/.xinitrc'`,  because it is so useful for editions that don't use displaymanager. 

I don't think autocorrect option is that dangerous, because it only commits changes after "did you mean?" - dialog. But as @jonathonf said, better to err on the side of caution.

**Comment by philmmanjaro** [comment]
@Chrysostomus: What is the current status on this one?

**Comment by Chrysostomus** [comment]
Works, has been used in some community editions for a long time now. The issues with qt terminals have been solved. If others agree, I feel that it is ready to be included

**Comment by Chrysostomus** [comment]
If you want some changes, it is fast and easy to implement them now.

**Comment by UWorkbench** [comment]
any news about this?

**Comment by philmmanjaro** [comment]
We already added some options for our maintainers to choose which console syntax they prefer. Most kept bash as default. You may find zsh within the openbox edition.

**Comment by Chrysostomus** [comment]
Awesome, bspwm and Bspwm-mate have always used zsh. Zsh is available as option for all editions in manjaro-architect. Recently we got a bug fixed that made manjaro-zsh-config more compatible with qt terminals, so there are no longer major issues with adopting it.

**Comment by codesardine** [comment]
In my newbie opinion if shell is meant to be more user friendly Fish is the way to go, it can be done without making it default for bash scripts, and I am going to implement that in a future version of webdad.

**Comment by Chrysostomus** [comment]
Have you tried manjaro-zsh-config? It implements most of the default features of fish (history substring search, autosuggestion, syntax highlighting). I think that it is more user friendly to have zsh with those features, because that way

- if user copies bash one liners from arch wiki etc, they work 
- if user wants to test portions of their own script, they can just copy-paste them to the terminal
- tab completion and spellchecking are slightly better in zsh

Anyway, fish is an selectable option manjaro-architect too. But I'm curious, what features of fish you feel make it better than manjaro-zsh-config?

**Comment by codesardine** [comment]
> if user copies bash one liners from arch wiki etc, they 
You are right about this.

 I was not aware manjaro-zsh config, I was referring if someone was starting working with terminal, Fish is easier to pick up, and more in intuitive to work with that normal bash.
